### PR TITLE
Updates in the data sources api reference

### PR DESCRIPTION
**Pull request by luisaFelixSalles**
Updates the DataSources API reference

- Type hints in all the functions ans properties of the class
- More examples and explanations

### CODE REVIEW

File: src/ansys/dpf/core/data_sources.py [hunk 1]
```diff
@@ -69,13 +77,21 @@ class DataSources:
...
     ['file.rst']
 
     """
 
-    def __init__(self, result_path=None, data_sources=None, server=None):
+    def __init__(
+        self,
+        result_path: Optional[Union[str, os.PathLike]] = None,
+        data_sources: Optional[dpf.DataSources] = None,
+        server: Optional[server_types.DpfServer] = None,
```
Comment (PProfizi) [inline src/ansys/dpf/core/data_sources.py:93]: Not `server_types.DpfServer`.
Either the `BaseServer` class, or a `Union` of `LegacyGrpcServer`, `GrpcServer`, and `InProcessServer`. I think the `Union` is best.

File: src/ansys/dpf/core/data_sources.py [hunk 2]
```diff
@@ -69,13 +77,21 @@ class DataSources:
...
     >>> my_data_sources.result_files
     ['file.rst']
 
     """
 
-    def __init__(self, result_path=None, data_sources=None, server=None):
+    def __init__(
+        self,
+        result_path: Optional[Union[str, os.PathLike]] = None,
+        data_sources: Optional[dpf.DataSources] = None,
```
Comment (PProfizi) [inline src/ansys/dpf/core/data_sources.py:92]: Not a `dpf.DataSources`.
A `Union` of either `int` for a memory address or `ansys.grpc.dpf.data_sources_pb2.DataSources` as in the original signature.

File: src/ansys/dpf/core/data_sources.py [hunk 3]
```diff
@@ -183,9 +260,9 @@ def guess_second_key(filepath: str) -> str:
         return new_key
 
     def set_domain_result_file_path(
-        self, path: Union[str, os.PathLike], domain_id: int, key: Union[str, None] = None
-    ):
-        """Add a result file path by domain.
+        self, path: Union[str, os.PathLike], domain_id: int, key: Optional[str] = None
```
Comment (PProfizi) [inline src/ansys/dpf/core/data_sources.py:263]:
```suggestion
        self, path: Union[str, os.PathLike], domain_id: int, key: Optional[str] = None
```
I think here `None` has to be listed as a valid type for the key argument.

File: src/ansys/dpf/core/data_sources.py [hunk 4]
```diff
@@ -395,12 +642,44 @@ def result_files(self):
...
+        When using an operator that requires data from a ``DataSources`` object, DPF must find
+        a corresponding entry to this call in its code. This entry is given
+        by the namespace, the file extension, and the operator name: ``namespace::key::operator_name``.
+
+        For example, if the results file comes from a MAPDL solver and has an '.rst' extension
+        and you want to get the displacement results in this file, DPF code will get the
+        correspondence: ``mapdl::rst::displacement``.
+
+        So, if you have an extension that is not
+        know by DPF you have to define its namespace. This function is mainly used when
```
Comment (PProfizi) [inline src/ansys/dpf/core/data_sources.py:655]:
```suggestion
        registered in DPF you have to define its namespace. This function is mainly used when
```

File: src/ansys/dpf/core/data_sources.py [hunk 5]
```diff
@@ -395,12 +642,44 @@ def result_files(self):
...
+        creating your own operators and plugins, or when you have a file with an unknown
+        namespace, but you know that it corresponds to certain solver.
```
Comment (PProfizi) [inline src/ansys/dpf/core/data_sources.py:657]:
```suggestion
        namespace, but you know that it corresponds to a given solver.
```

File: src/ansys/dpf/core/data_sources.py [hunk 6]
```diff
@@ -395,12 +642,44 @@ def result_files(self):
...
+
+        Parameters
+        ----------
+        result_key:
+            Extension of the file, which is used as a key for choosing the correct
+            plugin when a result is requested by an operator.
+        namespace:
+            Name of the correspondent solver
```
Comment (PProfizi) [inline src/ansys/dpf/core/data_sources.py:665]:
```suggestion
            Name of the corresponding solver
```

File: src/ansys/dpf/core/data_sources.py (hunk 3 above)
Comment (luisaFelixSalles) [reply src/ansys/dpf/core/data_sources.py:263]: I think when we use optional is exactly to omit the "None" in a union

File: src/ansys/dpf/core/data_sources.py [hunk 7]
```diff
@@ -272,11 +371,19 @@ def add_domain_file_path(self, filepath, key, domain_id):
...
         >>> from ansys.dpf import core as dpf
-        >>> data_sources = dpf.DataSources()
-        >>> data_sources.add_domain_file_path('/tmp/ds.dat', "dat", 1)
+        >>>
+        >>> # Create the DataSources object
+        >>> my_data_sources = dpf.DataSources()
+        >>> # Define the path where the main result data can be found and specify its domain
+        >>> my_data_sources.set_domain_result_file_path(path='/tmp/ds.cas', key='cas', domain_id=1)
+        >>> # Add the additional result data to the DataSources object and specify its domain
+        >>> my_data_sources.add_domain_file_path(path='/tmp/ds.dat', key="dat", domain_id=1)
```
Comment (luisaFelixSalles) [inline src/ansys/dpf/core/data_sources.py:386]:
```suggestion
        >>> my_data_sources.add_domain_file_path(filepath='/tmp/ds.dat', key="dat", domain_id=1)
```

File: src/ansys/dpf/core/data_sources.py [hunk 8]
```diff
@@ -44,9 +46,16 @@
 from ansys.dpf.core.check_version import version_requires
 from ansys.dpf.core import errors
 
+if TYPE_CHECKING:
```
Comment (PProfizi) [inline src/ansys/dpf/core/data_sources.py:49]:
```suggestion
if TYPE_CHECKING:  # pragma: no cover
```

File: src/ansys/dpf/core/data_sources.py [hunk 9]
```diff
@@ -376,6 +566,68 @@ def result_files(self):
...
+        >>> # Add the additional result file to the DataSources object
+        >>> my_data_sources.add_file_path(filepath='/tmp/ds.dat', key='dat')
+        >>> # Get the path to the main result file
+        >>> my_data_sources.result_files
+        ['/tmp/file.cas']
+
+        - If you added an upstream result file, it is not listed in the main ``DataSources`` object. You have to
+        check directly in the ``DataSources`` object created to define the upstream data.
+
+        Examples
```
Comment (PProfizi) [inline src/ansys/dpf/core/data_sources.py:592]: @luisaFelixSalles maybe the second `Examples` section title is the source of error

**Pull request closed without merging**